### ISSUE TITLE
False positive `missing-argument` with default values in function factories

### ISSUE THREAD

**Issue by honnorat** [astral-sh/ty#3863]
### Summary

Let's consider a function with a default argument value:

```python
def func(x: int = 0) -> int:
    return x + 1
```

When produced by a factory, `ty` incorrectly raises a `missing-argument` when this function is used without argument.

```python
from collections.abc import Callable

type Func = Callable[[int], int]


def func(x: int = 0) -> int:
    return x + 1

assert func(1) == 2
assert func() == 1


def make_func() -> Func:
    return func


func_gen = make_func()
assert func_gen(1) == 2
assert func_gen() == 1  # ty raises a missing-argument here
```

```shell
$ uvx ty check test_missing_arg.py
error[missing-argument]: No argument provided for required parameter 1
  --> test_missing_arg.py:19:8
   |
19 | assert func_gen() == 1  # ty raises a missing-argument here
   |        ^^^^^^^^^^
   |

Found 1 diagnostic
```
 
https://play.ty.dev/54902ff5-f116-4162-a31d-b8451762dc91

### Version

ty 0.0.54

**Comment by MeGaGiGaGon** [comment]
No other type checker (mypy, pyright, pyrefly) supports this pattern.

This happens because in a `Callable`, there is no way to annotate if a parameter has a default or not, so it is assumed that there is no default.

Instead, you can use a callable `Protocol`, which allow you to specify the calling expectations more precisely, like this:
https://play.ty.dev/46e2caf3-11df-4eb0-99e9-c38947d94ffc
```py
class Func(Protocol):
    def __call__(self, x: int = 0) -> int: ...
```
All type checkers (including ty) accept this version

**Comment by honnorat** [comment]
Thank you ! Now that you mention it, I think I've already fallen into that trap... 🫣
